File: firstWeek/SearchInRotatedSortedArray.java
class Solution {
    public int search(int[] nums, int target) {
        if (nums == null || nums.length <= 0) {
            return -1;
        }
        
        int low = 0;
        int high = nums.length - 1;
        while (low < high) {
            int mid = (low + high) >>> 1;
            if (nums[mid] > nums[high]) {
                low = mid + 1;
                continue;
            }
            high = mid;
        }
        
        int rotatedLength = low;
        low = 0;
        high = nums.length;
        while (low < high) {
            int mid = (low + high) >>> 1;
            int realMid = (mid + rotatedLength) % nums.length;
            if (nums[realMid] < target) {
                low = mid + 1;
                continue;
            }
            high = mid;
        }
        int index = (low + rotatedLength) % nums.length;
        return nums[index] == target ? index : -1;
    }
}

File: eightnWeek/search_in_rotated_sorted_array.py
class Solution:
    def search(self, nums: List[int], target: int) -> int:
        if len(nums) < 1:
            return -1

        low, high = 0, len(nums) - 1
        while low < high:
            mid = low + (high - low) // 2
            if nums[mid] > nums[high]:
                low = mid + 1
            else:
                high = mid

        rotated_len = low
        low, high = 0, len(nums)
        while low < high:
            mid = low + (high - low) // 2
            true_mid = (mid + rotated_len) % len(nums)
            if nums[true_mid] < target:
                low = mid + 1
            else:
                high = mid
        index = (low + rotated_len) % len(nums)
        return index if nums[index] == target else -1
